### PR TITLE
feat: update pvc accessmodes and resources when statefulset creates/u…

### PR DESCRIPTION
**Pull request by draveness**
**What type of PR is this?**

kind/feature

**What this PR does / why we need it**:

Update PVC accessmodes and storage when statefulset updates volumeClaimTemplates

Related to #57587

**Does this PR introduce a user-facing change?**:

```release-note
Update PVC accessmodes and storage when statefulset updates volumeClaimTemplates
```

### CODE REVIEW

**Comment by draveness** [comment]
Kindly ping @saad-ali @kow3ns for approval.

**Comment by timmycarr** [comment]
Hi! Friendly reminder from your release team: we are starting the code freeze for 1.15 tomorrow EOD. Just checking in to see if this is still planned for the 1.15 cycle?

**Comment by kow3ns** [comment]
We're looking at allowing for mutable volume claims, primarily to resize storage requests, but, currently, the API Machinery [rejects volume mutation](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/apps/validation/validation.go#L144).  i.e. you can't currently update volume claims so this can't work.

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 1]
```diff
@@ -189,11 +190,34 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
...
 			errs = append(errs, fmt.Errorf("Failed to retrieve PVC %s: %s", claim.Name, err))
 			spc.recordClaimEvent("create", set, pod, &claim, err)
+
+		default:
+			// Check resource requirements and accessmodes, update if necessary
+			claimNeedUpdated := false
+			if !volumeutil.AccessModesEqualSet(claim.Spec.AccessModes, currentClaim.Spec.AccessModes) {
+				currentClaim.Spec.AccessModes = claim.Spec.AccessModes
+				claimNeedUpdated = true
+			}
```
Comment (cwdsuzhou) [inline pkg/controller/statefulset/stateful_pod_control.go]: if claimNeedUpdated is true, update directly

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 2]
```diff
@@ -189,11 +190,34 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
...
+			claimStorage, claimOk := claim.Spec.Resources.Requests[v1.ResourceStorage]
+			currentStorage, currentOk := currentClaim.Spec.Resources.Requests[v1.ResourceStorage]
+			if claimOk && currentOk && claimStorage.Cmp(currentStorage) != 0 {
+				currentClaim.Spec.Resources.Requests[v1.ResourceStorage] = claimStorage
+				claimNeedUpdated = true
+			}
+
+			if claimNeedUpdated {
+				_, err := spc.client.CoreV1().PersistentVolumeClaims(claim.Namespace).Update(currentClaim)
+				if err != nil {
```
Comment (cwdsuzhou) [inline pkg/controller/statefulset/stateful_pod_control.go]: could we use patch

**Comment by kow3ns** [comment]
@cwdsuzhou As above, you can't actually mutate the volume claims, and there is a larger effort between SIG Apps and Storage to do so. This PR is not going to work.

**Comment by cwdsuzhou** [comment]
> @cwdsuzhou As above, you can't actually mutate the volume claims, and there is a larger effort between SIG Apps and Storage to do so. This PR is not going to work.

Thanks for your clarification.

**Comment by draveness** [comment]
> @cwdsuzhou As above, you can't actually mutate the volume claims, and there is a larger effort between SIG Apps and Storage to do so. This PR is not going to work.

Thanks @kow3ns, I'll close the PR. BTW, where could I find resources about mutable volume claims, I'm quite interested in it.

**Pull request closed without merging**